src/main/java/network/brightspots/rcv/ContestConfig.java

```diff
@@ -1025,22 +1038,11 @@ void setCandidateExclusionStatus(String candidateCode, boolean excluded) {
   }
 
   // perform pre-processing on candidates:
-  // 1) if there are any CDF input sources extract candidates names from them
-  // 2) build map of candidate ID to candidate name
-  // 3) generate tie-break ordering if needed
+  // 1) build map of candidate ID to candidate name
+  // 2) generate tie-break ordering if needed
+  // 3) add uwi candidate if needed
```

Comment:
Does this mean now we're requiring CDF configs to include candidates? Why did you have to make this change?

Reply:
@tarheel  Right - the main reason is, our tabulation architecture relies on the config file defining valid candidates for a contest to be tabulated.  This change treats all data sources "equally" in that respect.  It generalizes validation code across source types, and generalizes the feature set, which enables usage of candidate codes and excluding candidates from tabulation of CDF sources.

To be sure, simply using the embedded CDF candidate data is easier in that you don't have to bother putting it in the config (which I have spent a lot of time doing and not done yet) but I think -- and would like to get your feedback -- the better solution is to eventually provide a feature to extract the candidate data from the CDF into a config.

Reply:
Yeah, that all sounds fine to me. I agree that the ideal end state would be to require the data in the config, but make it easy to copy it over from the source.